### PR TITLE
include: drivers: display: verify if optional API is implemented

### PR DESCRIPTION
**Pull request by marcusfolkesson**
Pretty much every display driver contains several functions like this one:

static int otm8009a_set_contrast(const struct device *dev, uint8_t contrast)
{
	return -ENOTSUP;
}

Let the driver API return -ENOTSUP instead.

The pull request also contains a commit where one driver is cleaned up.
If the change is good for merge, I intend to cleanup the rest of the display drivers as well.

### CODE REVIEW

File: include/zephyr/drivers/display.h [hunk 1]
```diff
@@ -312,6 +324,10 @@ static inline int display_blanking_on(const struct device *dev)
 	struct display_driver_api *api =
 		(struct display_driver_api *)dev->api;
 
+	if (api->blanking_on) {
```
Comment (reatreada) [inline include/zephyr/drivers/display.h:327]:
```suggestion
	if (!api->blanking_on) {
```

File: include/zephyr/drivers/display.h [hunk 2]
```diff
@@ -331,6 +347,10 @@ static inline int display_blanking_off(const struct device *dev)
 	struct display_driver_api *api =
 		(struct display_driver_api *)dev->api;
 
+	if (api->blanking_off) {
```
Comment (reatreada) [inline include/zephyr/drivers/display.h:350]:
```suggestion
	if (!api->blanking_off) {
```

Comment (marcusfolkesson) [reply include/zephyr/drivers/display.h:350]: Outch. I switched from (!api->foo) to (api->foo == NULL) as "the last quick change" before commiting. Thank you for your keen eye.

File: include/zephyr/drivers/display.h [hunk 3]
```diff
@@ -247,6 +247,10 @@ static inline int display_write(const struct device *dev, const uint16_t x,
 	struct display_driver_api *api =
 		(struct display_driver_api *)dev->api;
 
+	if (api->write == NULL) {
+		return -ENOTSUP;
```
Comment (jfischer-no) [inline include/zephyr/drivers/display.h:251]: Someone tried this before, but somehow it got stuck. I am not sure what is right, ENOTSUP or ENOSYS, https://docs.zephyrproject.org/latest/develop/api/design_guidelines.html is not clear at all. e.g. https://github.com/zephyrproject-rtos/zephyr/blob/673f38f3d1ccbb8d66edce195805deae439272d4/include/zephyr/drivers/auxdisplay.h#L327-L329 uses ENOSYS.
cc @nordicjm @kartben

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: include/zephyr/drivers/display.h (hunk 3 above)
Comment (nordicjm) [reply include/zephyr/drivers/display.h:251]: `ENOSYS` is correct here, from that page:
```
- APIs that are supported but not implemented shall return ``-ENOSYS``.

- Optional APIs that are not supported by the hardware should be implemented and
  the return code in this case shall be ``-ENOTSUP``.

- When an API is implemented, but the particular combination of options
  requested in the call cannot be satisfied by the implementation the call shall
  return ``-ENOTSUP``. (For example, a request for a level-triggered GPIO interrupt on
  hardware that supports only edge-triggered interrupts)
```

This is the top one. The second one applies in a driver for a specific peripheral e.g. display driver for hn238 but not the overall display.h file

Comment (marcusfolkesson) [reply include/zephyr/drivers/display.h]: I'll change to -ENOSYS.

File: include/zephyr/drivers/display.h [hunk 4]
```diff
@@ -247,6 +247,10 @@ static inline int display_write(const struct device *dev, const uint16_t x,
 	struct display_driver_api *api =
 		(struct display_driver_api *)dev->api;
 
+	if (api->write == NULL) {
```
Comment (jfischer-no) [inline include/zephyr/drivers/display.h:250]: Please add 

`* @retval -ENOSYS if not implemented.`

where it make sense.

File: include/zephyr/drivers/display.h [hunk 5]
```diff
@@ -247,6 +247,10 @@ static inline int display_write(const struct device *dev, const uint16_t x,
 	struct display_driver_api *api =
```
Comment (jfischer-no) [inline include/zephyr/drivers/display.h:247]: Please revise you commit title/message, maybe something like:
`include: drivers: display: verify if optional API is implemented`.
IIRC "sanity check" is a bit too vulgar.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: include/zephyr/drivers/display.h (hunk 2 above)
Comment (GeorgeCGV) [reply include/zephyr/drivers/display.h:354]: Missing:

```suggestion
	if (api->blanking_off == NULL) {
```

Comment (GeorgeCGV) [reply include/zephyr/drivers/display.h:354]: Blanking off condition is not handled, the rest LGTM.

**Comment by jfischer-no** [comment]
@marcusfolkesson Note: I will mark conversations as resolved if they are actually resolved.

File: include/zephyr/drivers/display.h [hunk 6]
```diff
@@ -364,12 +393,17 @@ static inline int display_set_brightness(const struct device *dev,
...
  *
  * @retval 0 on success else negative errno code.
+ * @retval -ENOSYS if not implemented.
  */
 static inline int display_set_contrast(const struct device *dev, uint8_t contrast)
 {
 	struct display_driver_api *api =
 		(struct display_driver_api *)dev->api;
 
+	if (api->set_brightness == NULL) {
```
Comment (jfischer-no) [inline include/zephyr/drivers/display.h:403]: `
	if (api->set_contrast == NULL) {
`

File: include/zephyr/drivers/display.h [hunk 7]
```diff
@@ -386,6 +420,10 @@ static inline void display_get_capabilities(const struct device *dev,
 	struct display_driver_api *api =
 		(struct display_driver_api *)dev->api;
 
+	if (api->get_capabilities == NULL) {
+		return;
+	}
+
```
Comment (jfischer-no) [inline include/zephyr/drivers/display.h:426]: That one cannot be optional but must be mandatory API.

File: include/zephyr/drivers/display.h [hunk 8]
```diff
@@ -247,6 +248,10 @@ static inline int display_write(const struct device *dev, const uint16_t x,
 	struct display_driver_api *api =
 		(struct display_driver_api *)dev->api;
 
+	if (api->write == NULL) {
+		return -ENOSYS;
+	}
+
```
Comment (jfischer-no) [inline include/zephyr/drivers/display.h:254]: Actually, display_write() is mandatory API.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

**Review by jfischer-no — APPROVED** [review]
Thanks

**Review by nordicjm — APPROVED** [review]

**Pull request merged**